iib/workers/tasks/build_regenerate_bundle.py

```diff
@@ -0,0 +1,356 @@
+            if operator_csv.has_related_image_envs():
+                raise IIBError(
+                    f'The ClusterServiceVersion file {csv_file_name} has entries in '
+                    'spec.relatedImages and one or more containers have RELATED_IMAGE_* '
+                    'environment variables set. This is not allowed for bundles regenerated with '
+                    'IIB.'
+                )
+            log.debug(
+                'Skipping pinning since the ClusterServiceVersion file %s has entries in '
+                'spec.relatedImages',
+                csv_file_name,
+            )
+            continue
+
+        operator_csvs.append(operator_csv)
+
+        for pullspec in operator_csv.get_pullspecs():
+            found_pullspecs.add(pullspec)
+
+    conf = get_worker_config()
+    registry_replacements = (
+        conf['iib_organization_customizations']
+        .get(organization, {})
+        .get('registry_replacements', {})
+    )
+
+    # Resolve pull specs to container image digests
+    replacement_pullspecs = {}
+    for pullspec in found_pullspecs:
+        replacement_needed = False
+        new_pullspec = ImageName.parse(pullspec.to_str())
+
+        if not pinned_by_iib:
+            # Resolve the image only if it has not already been processed by IIB. This
+            # helps making sure the pullspec is valid
+            resolved_image = ImageName.parse(get_resolved_image(pullspec.to_str()))
+
+            # If the tag is in the format "<algorithm>:<checksum>", the image is already pinned.
+            # Otherwise, always pin it to a digest.
+            if ':' not in ImageName.parse(pullspec).tag:
+                log.debug('%s will be pinned to %s', pullspec, resolved_image.to_str())
+                new_pullspec = resolved_image
+                replacement_needed = True
+                labels['com.redhat.iib.pinned'] = 'true'
+
+        # Apply registry modifications
+        new_registry = registry_replacements.get(new_pullspec.registry)
+        if new_registry:
+            replacement_needed = True
+            new_pullspec.registry = new_registry
+
+        if replacement_needed:
+            log.debug('%s will be replaced with %s', pullspec, new_pullspec.to_str())
+            replacement_pullspecs[pullspec] = new_pullspec
+
+    # Apply modifications to the operator bundle image metadata
+    for operator_csv in operator_csvs:
+        csv_file_name = os.path.basename(operator_csv.path)
+        log.info('Replacing the pull specifications on %s', csv_file_name)
+        operator_csv.replace_pullspecs_everywhere(replacement_pullspecs)
+
+        log.info('Setting spec.relatedImages on %s', csv_file_name)
+        operator_csv.set_related_images()
+
+        operator_csv.dump()
+
+    if organization:
+        _adjust_csv_annotations(operator_manifest.files, package_name, organization)
+
+    return labels
+
+
+def _adjust_csv_annotations(operator_csvs, package_name, organization):
```

Comment:
Do we want to get rid of `_` at the beginning of the name of the function?

Reply:
I don't think so because it's still a private function being used only in the same module.